consensus/src/quorum_store/batch_generator.rs

```diff
@@ -0,0 +1,388 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::block_storage::BlockReader;
+use crate::quorum_store::batch_coordinator::BatchCoordinatorCommand;
+use crate::quorum_store::counters;
+use crate::quorum_store::quorum_store_db::BatchIdDB;
+use crate::quorum_store::types::BatchId;
+use crate::quorum_store::utils::{BatchBuilder, MempoolProxy, RoundExpirations};
+use aptos_consensus_types::common::{Round, TransactionSummary};
+use aptos_consensus_types::proof_of_store::{LogicalTime, ProofOfStore};
+use aptos_logger::prelude::*;
+use aptos_mempool::QuorumStoreRequest;
+use futures::future::BoxFuture;
+use futures::stream::FuturesUnordered;
+use futures::StreamExt;
+use futures_channel::mpsc::Sender;
+use futures_channel::oneshot;
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+use tokio::{sync::mpsc::Sender as TokioSender, time::Interval};
+
+type ProofCompletedChannel = oneshot::Receiver<Result<(ProofOfStore, BatchId), ProofError>>;
+
+#[derive(Debug)]
+pub enum BatchGeneratorCommand {
+    CommitNotification(LogicalTime),
+    Shutdown(tokio::sync::oneshot::Sender<()>),
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum ProofError {
+    Timeout(BatchId),
+}
+
+pub struct BatchGenerator {
+    db: Arc<dyn BatchIdDB>,
+    mempool_proxy: MempoolProxy,
+    batch_coordinator_tx: TokioSender<BatchCoordinatorCommand>,
+    batches_in_progress: HashMap<BatchId, Vec<TransactionSummary>>,
+    batch_expirations: RoundExpirations<BatchId>,
+    batch_builder: BatchBuilder,
+    latest_logical_time: LogicalTime,
+    mempool_txn_pull_max_count: u64,
+    mempool_txn_pull_max_bytes: u64,
+    batch_expiry_round_gap_when_init: Round,
+    end_batch_ms: u128,
+    last_end_batch_time: Instant,
+    // temp variable for debug the txn seq number too new issue
+    max_batch_id: u64,
+    last_batch_id: u64,
+    remaining_proof_num: usize,
```

Comment:
Back pressure mechanism is broken. remaining_proof_num is currently updated in proof_manager but used in batch_generater. (previously they both were part of the wrapper).